File: python/openforge_sdk/ollama.py
from __future__ import annotations

import json
from typing import Any, TypeVar, cast

from .client import OpenForgeClient, OpenForgeError

T = TypeVar("T")


def _decode_mcp(result: dict[str, Any]) -> Any:
    if result.get("isError"):
        messages = [
            item.get("text", "")
            for item in result.get("content", [])
            if isinstance(item, dict) and item.get("type") == "text"
        ]
        raise OpenForgeError("\n".join(message for message in messages if message) or "Ollama MCP call failed")
    if "structuredContent" in result:
        return result["structuredContent"]
    for item in result.get("content", []):
        if isinstance(item, dict) and item.get("type") == "text" and isinstance(item.get("text"), str):
            return json.loads(item["text"])
    raise OpenForgeError("Ollama MCP response contained no structured result")


async def list_ollama_models(client: OpenForgeClient) -> dict[str, Any]:
    return cast(dict[str, Any], _decode_mcp(await client.mcp_call_tool("ollama", "list_models", {})))


async def preflight_ollama_model(
    client: OpenForgeClient, model: str, context_tokens: int
) -> dict[str, Any]:
    return cast(
        dict[str, Any],
        _decode_mcp(
            await client.mcp_call_tool(
                "ollama",
                "preflight_model",
                {"model": model, "context_tokens": context_tokens},
            )
        ),
    )


async def show_ollama_model(client: OpenForgeClient, model: str) -> dict[str, Any]:
    return cast(
        dict[str, Any],
        _decode_mcp(await client.mcp_call_tool("ollama", "show_model", {"model": model})),
    )


async def pull_ollama_model(client: OpenForgeClient, model: str) -> dict[str, Any]:
    return cast(
        dict[str, Any],
        _decode_mcp(await client.mcp_call_tool("ollama", "pull_model", {"model": model})),
    )

File: packages/sdk-typescript/src/ollama.ts
import type { OpenForgeClient } from "./client.js";

export interface OllamaModelInfo {
  name: string;
  model: string;
  size: number;
  digest: string;
  modified_at: string;
  details: Record<string, unknown>;
}

export interface OllamaInventory {
  models: OllamaModelInfo[];
  available_memory_bytes: number | null;
}

export interface OllamaPreflight {
  ready: boolean;
  installed: boolean;
  model: string;
  context_tokens: number;
  model_size_bytes?: number;
  available_memory_bytes: number | null;
  estimated_required_memory_bytes: number | null;
  message: string;
}

interface McpResult {
  content: unknown[];
  isError: boolean;
  structuredContent?: unknown;
}

function decodeMcp<T>(result: McpResult): T {
  if (result.isError) {
    const message = result.content
      .filter((item): item is { type: string; text: string } =>
        typeof item === "object" && item !== null &&
        "type" in item && (item as { type?: unknown }).type === "text" &&
        "text" in item && typeof (item as { text?: unknown }).text === "string")
      .map((item) => item.text)
      .join("\n");
    throw new Error(message || "Ollama MCP call failed");
  }
  if (result.structuredContent !== undefined) {
    return result.structuredContent as T;
  }
  for (const item of result.content) {
    if (
      typeof item === "object" && item !== null &&
      "type" in item && (item as { type?: unknown }).type === "text" &&
      "text" in item && typeof (item as { text?: unknown }).text === "string"
    ) {
      return JSON.parse((item as { text: string }).text) as T;
    }
  }
  throw new Error("Ollama MCP response contained no structured result");
}

export async function listOllamaModels(client: OpenForgeClient): Promise<OllamaInventory> {
  return decodeMcp<OllamaInventory>(
    await client.mcpCallTool("ollama", "list_models", {}),
  );
}

export async function preflightOllamaModel(
  client: OpenForgeClient,
  model: string,
  contextTokens: number,
): Promise<OllamaPreflight> {
  return decodeMcp<OllamaPreflight>(
    await client.mcpCallTool("ollama", "preflight_model", {
      model,
      context_tokens: contextTokens,
    }),
  );
}

export async function showOllamaModel(
  client: OpenForgeClient,
  model: string,
): Promise<Record<string, unknown>> {
  return decodeMcp<Record<string, unknown>>(
    await client.mcpCallTool("ollama", "show_model", { model }),
  );
}

export async function pullOllamaModel(
  client: OpenForgeClient,
  model: string,
): Promise<Record<string, unknown>> {
  return decodeMcp<Record<string, unknown>>(
    await client.mcpCallTool("ollama", "pull_model", { model }),
  );
}
